nav2_util/include/nav2_util/simple_action_server.hpp

```diff
@@ -137,22 +165,35 @@ class SimpleActionServer
   {
     std::lock_guard<std::recursive_mutex> lock(update_mutex_);
     server_active_ = true;
+    stop_execution_ = false;
   }
 
   void deactivate()
   {
-    std::lock_guard<std::recursive_mutex> lock(update_mutex_);
-    server_active_ = false;
+    debug_msg("Deactivating...");
 
-    if (is_active(current_handle_)) {
-      warn_msg("Taking action server to deactive state with an active goal.");
+    {
+      std::lock_guard<std::recursive_mutex> lock(update_mutex_);
+      server_active_ = false;
+      stop_execution_ = true;
     }
 
-    if (is_active(pending_handle_)) {
-      warn_msg("Taking action server to deactive state with a pending preemption.");
+    if (is_running()) {
+      warn_msg("Requested to deactivate server but goal is still executing."
+        " Should check if action server is running before deactivating.");
     }
 
-    terminate_goals();
+    using namespace std::chrono_literals;  //NOLINT
+    while (execution_future_.wait_for(100ms) != std::future_status::ready) {
+      info_msg("Waiting for async process to finish.");
+    }
+    debug_msg("Deactivation completed.");
+  }
+
+  bool is_running()
+  {
+    using namespace std::chrono_literals;  //NOLINT
+    return execution_future_.wait_for(0ms) == std::future_status::ready ? true : false;
```

Comment:
`return execution_future_.wait_for(0ms) == std::future_status::ready;` is the same thing :wink:

Reply:
:smiley: